### PR TITLE
staking miner: Check the queue one last time before submission 

### PR DESCRIPTION
**Pull request by niklasad1**
Builds on https://github.com/paritytech/polkadot/pull/4716

Closing #3740

### CODE REVIEW

File: utils/staking-miner/src/monitor.rs
Comment (niklasad1) [inline utils/staking-miner/src/monitor.rs:78]: we could make a batch request for this instead, not sure what's best trade-off TBH.

File: utils/staking-miner/src/main.rs
Comment (kianenigma) [inline utils/staking-miner/src/main.rs:315]:
```suggestion
		let res = if s == "--if-leading" {
```

File: utils/staking-miner/src/main.rs
Comment (kianenigma) [inline utils/staking-miner/src/main.rs:317]:
```suggestion
		} else if s == "--always" {
```

File: utils/staking-miner/src/main.rs
Comment (kianenigma) [inline utils/staking-miner/src/main.rs:321]: this should parse something like `--if-better-than 10` (10 would be passed to `Perbill::from_precent`)

File: utils/staking-miner/src/monitor.rs
Comment (kianenigma) [inline utils/staking-miner/src/monitor.rs:70]: hard-coding `EPM_MODULE_PREFIX` is wrong, since it is coming from the name of the pallet in `construct_runtime`, and it could be different. 

You can however get this from `T`: https://paritytech.github.io/substrate/master/frame_support/traits/trait.PalletInfo.html#tymethod.name

Comment (kianenigma) [inline utils/staking-miner/src/monitor.rs:70]: Also, as discussed, you can make `SignedSubmissionIndices` public.

File: utils/staking-miner/src/monitor.rs
Comment (kianenigma) [inline utils/staking-miner/src/monitor.rs:84]: I would wait until your other PR is merged (which you can do ASAP since it is all approved now) to rewrite this stuff with the new PRC helpers? I think the code would be cleaner.

File: utils/staking-miner/src/monitor.rs
Comment (kianenigma) [inline utils/staking-miner/src/monitor.rs:100]: This one needs a rewrite, there must have been a misunderstanding about it. 

We want to compare ourselves with two queues.

1. The stuff stored in `SignedSubmissionsMap`. These are unverified solutions, and are sorted based on score. We want to scan these solutions, and then react to them based on `SubmissionStrategy`. Recall that the reason that we might want to do something like `AlwaysSubmit` is that these solutions are unverified, and they might have a false high claimed score. 

2. Then you have `QueuedSolution`. This is a single verified solution, that is the current best solution. In practice, we don't need to check this, since it is always empty during the signed phase (it is only set at the end of the signed phase). But since you already have the code, we should just basically treat it always as `OnlySubmitIfLeading`. As in, we check this, if it exists and if it has a better score than us, then we don't submit. Otherwise, we do.

**Review by kianenigma — CHANGES_REQUESTED** [review]
- Some misunderstanding about the queues that need to be checked.
- Can use `PalletInfo`.
- Can use the nice RPC helpers from the other PR.

File: utils/staking-miner/src/monitor.rs
Comment (niklasad1) [reply utils/staking-miner/src/monitor.rs]: I had some troubles to actually use the `StorageValue trait` here because of some trait bounds that are missing from the `Runtime`.

```bash
let xx = StorageValue::<EPM::SignedSubmissions::<Runtime>>::hashed_key();

    = note: required because of the requirements on the impl of `FullEncode` for `SignedSubmissions<kusama_runtime::Runtime>`
    = note: required because of the requirements on the impl of `FullCodec` for `SignedSubmissions<kusama_runtime::Runtime>`
note: required by a bound in `hashed_key`
```

File: utils/staking-miner/src/main.rs
Comment (niklasad1) [inline utils/staking-miner/src/main.rs:330]: this is a little but weird/controversial one need to passing this a string such as:

```$ target/release/staking-miner --uri ws://localhost:9999 --seed //Alice monitor --submission-strategy "percent-better 1" phrag-mms
```

we could just take a number but :point_up: is a bit more readable

File: utils/staking-miner/src/monitor.rs
Comment (niklasad1) [inline utils/staking-miner/src/monitor.rs:244]: //cc @emostov now this these calls are performed in parallel however the code is a bit nasty with the clones

File: utils/staking-miner/src/monitor.rs
Comment (niklasad1) [inline utils/staking-miner/src/monitor.rs:243]:
```suggestion
			// Run the calls in parallel and return once all has completed or any failed.
```

File: utils/staking-miner/src/monitor.rs
Comment (niklasad1) [inline utils/staking-miner/src/monitor.rs:307]:
```suggestion
			// Run the calls in parallel and return once all has completed or any failed.
```

**Comment by niklasad1** [comment]
@kianenigma can you take another look at this? I tried to address your grumbles and I hope I got it right this time

File: utils/staking-miner/src/main.rs
Comment (emostov) [inline utils/staking-miner/src/main.rs:361]:
```suggestion
	/// Submission strategy to use. Ex: `--submission-strategy "percent-better 1"`.
```
Not sure what the convention is for clap CLIs, but it would be nice if the user could get an example arg for percent-better since its a bit tricky

File: utils/staking-miner/src/monitor.rs
Comment (emostov) [inline utils/staking-miner/src/monitor.rs:107]: Not sure how all the underlying code works, but this looks like it will linearly traverse all the indices to get to the last one. `BTreeMap::last_key_value` presumably can do this with just logarithmic complexity. So this could be changed to `if let Some(curr_max_score) = indices.into_iter().last_key_value().map(|(s, _)| s)`

File: utils/staking-miner/src/monitor.rs
Comment (emostov) [inline utils/staking-miner/src/monitor.rs:120]: I don't think this is neccesary at all. The first time a `ReadySolution` is available is when we finalize the signed phase, which is at the very end. We should already check that the we are in the signed phase, which should implicitly ensure the `ReadySolution` does not exist. So I think you can remove this function

File: utils/staking-miner/src/monitor.rs
Comment (emostov) [inline utils/staking-miner/src/monitor.rs:212]: Just out of curiosity, why a function instead of a closue? Is it a limitation of the `async` keyword?

File: utils/staking-miner/src/monitor.rs
Comment (emostov) [inline utils/staking-miner/src/monitor.rs:227]: imo, if possible we should try and put this in the try join and then have a special match case for the `check_versions` error where we do `tx.send(err.into()); return;`

File: utils/staking-miner/src/monitor.rs
Comment (emostov) [reply utils/staking-miner/src/monitor.rs:85]: > But since you already have the code, we should just basically treat it always as OnlySubmitIfLeading. As in, we check this, if it exists and if it has a better score than us, then we don't submit. Otherwise, we do.

Not sure if I am reading this correctly, but I think there is no need to check the queued solution - we could instead check the phase. AFAICT in the current code its an invariant that if we set a QueuedSolution, then by the end of the block the phase will not be Signed

File: utils/staking-miner/src/monitor.rs
Comment (emostov) [inline utils/staking-miner/src/monitor.rs:310]:
```suggestion
```

File: utils/staking-miner/src/monitor.rs
Comment (emostov) [inline utils/staking-miner/src/monitor.rs:315]:
```suggestion
```

File: utils/staking-miner/src/monitor.rs
Comment (emostov) [reply utils/staking-miner/src/monitor.rs:120]: If we need to check the phase we can just query `EPM::CurrentPhase::<Runtime>`, which should be a lot cheaper then trying to query for a full solution

File: utils/staking-miner/src/monitor.rs
Comment (emostov) [inline utils/staking-miner/src/monitor.rs:91]: 👍

File: utils/staking-miner/src/monitor.rs
Comment (niklasad1) [reply utils/staking-miner/src/monitor.rs:107]: The `last_first APIs` are not stabilized yet so I can't use them yet but those seems to be optimized [see this](https://github.com/rust-lang/rust/issues/62924)

Since `BTreeMap::iter` implements `DoubleEndedIterator` this should be faster than linear AFAIU.

File: utils/staking-miner/src/monitor.rs
Comment (niklasad1) [reply utils/staking-miner/src/monitor.rs:227]: the issue is that  `try_join` will return as soon one future fails i.e, returns `Err()`

so this might fail but something else always get first and the main task will never get notified that 
the runtime version is outdated and it will try to spawn new tasks that always will fail..

that's the reason why it did it like this but I guess I could do `join(check_version, try_join(a, b, c))` instead

File: utils/staking-miner/src/main.rs
Comment (niklasad1) [reply utils/staking-miner/src/main.rs:361]: yeah, this is a bit weird maybe using an integer will work better with the CLI help menu.

File: utils/staking-miner/src/monitor.rs
Comment (niklasad1) [reply utils/staking-miner/src/monitor.rs:212]: async closures are not supported in rust yet, so that's why it's a separate function.

This could be a helper method because `StakingMinerError` needs `T` from the `Runtime` I made it an internal helper.. :(

File: utils/staking-miner/src/main.rs
Comment (kianenigma) [reply utils/staking-miner/src/main.rs:330]: I think this is fine 👍

File: utils/staking-miner/src/main.rs
Comment (kianenigma) [inline utils/staking-miner/src/main.rs:632]: no -- we just spoke of this, we only need to check this strictly in `monitor` in other commands it can be a mismatch.

File: utils/staking-miner/src/monitor.rs
Comment (kianenigma) [inline utils/staking-miner/src/monitor.rs:67]: I kinda wanted to rewrite this with `.any()` but I realized async code in closures it still a PITA right?

File: utils/staking-miner/src/monitor.rs
Comment (kianenigma) [inline utils/staking-miner/src/monitor.rs:109]:
```suggestion
			return Err(Error::StrategyNotSatisfied)
```

File: utils/staking-miner/src/monitor.rs
Comment (kianenigma) [reply utils/staking-miner/src/monitor.rs:120]: yeah, in the previous round of review I opted to keep this part, even though it should never be needed. Let's do what @emostov suggests for now: remove this check for simplicity.

File: utils/staking-miner/src/monitor.rs
Comment (kianenigma) [reply utils/staking-miner/src/monitor.rs:212]: AFAIK `fn` and `const` inside a body of the function are identical to being places in global scope. So, unlike closures, they do not capture anything from their scope.

File: utils/staking-miner/src/monitor.rs
Comment (niklasad1) [reply utils/staking-miner/src/monitor.rs:227]: I realized it would be slightly better to have a runtime version subscription here instead of polling `runtime_version` all over the place, when it received a new version -> quit all

File: utils/staking-miner/src/monitor.rs
Comment (niklasad1) [reply utils/staking-miner/src/monitor.rs:67]: it's possible but then we need convert the futures to a stream or batch them together.

Something like [FuturesOrdered](https://docs.rs/futures/latest/futures/stream/struct.FuturesUnordered.html#) or JSON-RPC batch request might make this is little bit nicer to read.

**Comment by niklasad1** [comment]
> LGTM, as a bonus test, you can try and run this against wesntend or kusama (you probably need to "fake" the linked runtime spec version).

Running it right now, I'll keep you posted.

File: utils/staking-miner/src/monitor.rs
Comment (niklasad1) [reply utils/staking-miner/src/monitor.rs:225]: I'm going to revert the `create_election_ext` because it increases network usage significantly even if the we drop the requests the server will answer it anyway.

In addition we only need to fetch these in signed phase anyway still no performance issues what I'm aware of anyway.

File: utils/staking-miner/src/monitor.rs
Comment (niklasad1) [reply utils/staking-miner/src/monitor.rs:202]: unfortunately I ran into https://github.com/paritytech/substrate/issues/10910 occasionally when I tested it

**Comment by niklasad1** [comment]
@emostov can you take another look at this as well?

File: utils/staking-miner/src/monitor.rs
Comment (emostov) [inline utils/staking-miner/src/monitor.rs:211]: I assume we do this here to prevent heavy network usage from `create_election_ext` and then again right before submitting to make sure its not pointless to submit? 

I think its fine for now, but in the future I think we should look into only doing it right before submitting. Even if it ends up with a lot of wasted network usage it should be fine as long as we are not executing this function righter _before_ the signed phase, since that is when we want to be really fast to get the solution in. But if there fear is that this will be executing at the end of the sign phase I think its fine because we will have awhile until the next signed phase, so the extra network usage won't slow itself down. (Assuming its just traffic to a node on the local network)

**Review by emostov — APPROVED** [review]

**Comment by niklasad1** [comment]
bot merge

**Pull request merged**
